core/src/main/java/com/auth0/authentication/api/internal/SimpleRequest.java

```diff
@@ -41,52 +41,65 @@
 import java.io.InputStream;
 import java.util.Map;
 
-class SimpleRequest<T> extends BaseRequest<T> implements Request<T>, ParameterizableRequest<T>, Callback {
+class SimpleRequest<T> extends BaseRequest<T> implements ParameterizableRequest<T>, Callback {
 
-    private final ObjectReader errorReader;
     private final String method;
 
     public SimpleRequest(HttpUrl url, OkHttpClient client, ObjectMapper mapper, String httpMethod, Class<T> clazz) {
-        super(url, client, mapper.reader(clazz), mapper.writer());
-        this.errorReader = mapper.reader(new TypeReference<Map<String, Object>>() {});
+        super(url, client, mapper.reader(clazz), mapper.reader(new TypeReference<Map<String, Object>>() {}), mapper.writer());
         this.method = httpMethod;
     }
 
     public SimpleRequest(HttpUrl url, OkHttpClient client, ObjectMapper mapper, String httpMethod) {
-        super(url, client, mapper.reader(new TypeReference<Map<String, Object>>() {}), mapper.writer());
-        this.errorReader = mapper.reader(new TypeReference<Map<String, Object>>() {
-        });
+        super(url, client, mapper.reader(new TypeReference<Map<String, Object>>() {}), mapper.reader(new TypeReference<Map<String, Object>>() {}), mapper.writer());
         this.method = httpMethod;
     }
 
     @Override
     public void onResponse(Response response) throws IOException {
-        final InputStream byteStream = response.body().byteStream();
         if (!response.isSuccessful()) {
-            Throwable throwable;
-            try {
-                Map<String, Object> payload = errorReader.readValue(byteStream);
-                throwable = new APIClientException("Request failed with response " + payload, response.code(), payload);
-            } catch (IOException e) {
-                throwable = new APIClientException("Request failed", response.code(), null);
-            }
-            postOnFailure(throwable);
+            APIException exception = parseUnsuccessfulResponse(response);
+            postOnFailure(exception);
             return;
         }
 
         try {
+            final InputStream byteStream = response.body().byteStream();
             T payload = getReader().readValue(byteStream);
             postOnSuccess(payload);
-        } catch (IOException e) {
-            postOnFailure(new APIClientException("Request failed", response.code(), null));
+        } catch (Exception e) {
```

Comment:
If we catch only `IOException` does the compiler complains?. Also the JSON parsing raises something that extends `IOException`